EIPS/eip-3224.md

```diff
@@ -0,0 +1,465 @@
+---
+eip: 3224
+title: Described Data and Described Transactions
+author: Richard Moore (@ricmoo), Nick Johnson (@arachnid)
+discussions-to: https://github.com/ethereum/EIPs/issues/3225
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-01-23
+requires: 191
+---
+
+## Simple Summary
+
+A technique for contract authors to enable wallets to provide
+a human-readable description of what a given contract
+interaction (via a transaction or signed message) will perform.
+
+
+## Abstract
+
+Human-readable descriptions for machine executable operations,
+described in higher level machine readable data, so that wallets
+can provide meaningful feedback to the user describing the
+action the user is about to perform.
+
+
+## Motivation
+
+When using an Ethereum Wallet (e.g. MetaMask, Clef, Hardware
+Wallets) users must accept and authorize signing messages or
+sending transactions.
+
+Due to the complexity of Ethereum transactions, wallets are very
+limitd in their ability to provide insight into the contents of
+transactions user are approving; outside special-cased support
+for common transactions such as ERC20 transfers, this often amounts
+to asking the user to sign an opaque blob of binary data.
+
+This EIP presents a method for dapp developers to enable a more
+comfortable user experience by providing wallets with a means
+to generate a better description about what the contract claims
+will happen.
+
+It does not address malicious contracts which wish to lie, it
+only addresses honest contracts that want to make their user's
+life better. We believe that this is a reasonable security model,
+as transaction descriptions can be audited at the same time as
+contract code, allowing auditors and code reviewers to check that
+transaction descriptions are accurate as part of their review.
+
+
+## Specification
+
+The **description string** and **described data** are generated
+simultaneously by evaluating **describer** bytecode, passed
+**describedInput** as the calldata with the encoded method:
```

Comment:
Most of these names don't align with parameter names in the Solidity below, which makes it really difficult to tell what is being talked about.  Recommend following a consistent naming scheme throughout the document.

Reply:
You mean I should rename the signature to:

```
function eipXXXDescribe(bytes describer_input) view returns (string description_string, bytes described_data);
```

?

I can make that change now.

Reply:
I've updated this section.

Reply:
Yes, though I still don't know what `describer` is.

Reply:
The describer is just the contract itself. I can just change the name to contract? Originally this was more abstract, but we decided it made more sense to reduce the scope to a contract doing this itself (as a delegatecall can be used to outsource to another contract rather than returning a separated address to operate as the describer).